webserver/cWebem.cpp

```diff
@@ -1157,6 +1157,115 @@ namespace http {
 			m_session_clean_timer.async_wait([this](auto &&) { CleanSessions(); });
 		}
 
+		bool cWebem::findRealHostBehindProxies(const request &req, std::string &realhost)
+		{
+			// Checking for 3 possible headers (in order of handling)
+			// "Forwarded"	RFC7239  (https://www.rfc-editor.org/rfc/rfc7239)
+			// "X-Forwarded-For" The defacto standard header used by many web/proxy servers (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For)
+			// "X-Real-IP"	The (old) default header used by NGINX  (http://nginx.org/en/docs/http/ngx_http_realip_module.html#real_ip_header)
+			//
+			// These headers can occur multiple times, so need to be 'squashed' together
+			// And a single line can contain multiple (comma separated) values in order
+
+			bool bStatus = true;
+			std::vector<std::string> headers;
+			std::vector<std::string> hosts;
+
+			if (sumProxyHeader("Forwarded", req, headers))
+			{	// We found one or more Forwarded headers that need to be processed into a list of Hosts
+				if (!parseForwardedProxyHeader(headers, hosts))
+				{
+					bStatus = false;
+				}
+			}
+			else if (sumProxyHeader("X-Forwarded-For", req, headers))
+			{	// We found one or more X-Forwarded-For headers that need to be processed into a list of Hosts
+				if (!parseProxyHeader(headers, hosts))
+				{
+					bStatus = false;
+				}
+			}
+			else if (sumProxyHeader("X-Real-IP", req, headers))
+			{	// We found one or more X-Real-IP headers that need to be processed into a list of Hosts
+				if (!parseProxyHeader(headers, hosts))
+				{
+					bStatus = false;
+				}
+			}
+
+			if (hosts.size() >= 1)
+			{
+				realhost = hosts[0];	// Even if we found a chain of hosts, we always use the first (= origin)
+			}
+
+			return bStatus;
+		}
+
+		bool cWebem::sumProxyHeader(const std::string sHeader, const request &req, std::vector<std::string> &vHeaderLines)
+		{
+			uint8_t iFound = 0;
+
+			for (const auto &header : req.headers)
+			{
+				if (sHeader.compare(header.name) == 0)
+				{
+					iFound++;
+					vHeaderLines.push_back(header.value);
+				}
+			}
+
+			return (iFound > 0);
+		}
+
+		bool cWebem::parseProxyHeader(std::vector<std::string> &vHeaderLines, std::vector<std::string> &vHosts)
+		{
+			uint8_t iFound = 0;
+
+			for (const auto sLine : vHeaderLines)
+			{
+				std::vector<std::string> vLineParts;
+				StringSplit(sLine, ",", vLineParts);
+				for (std::string sPart : vLineParts)
+				{
+					iFound++;
+					stdstring_trimws(sPart);
+					vHosts.push_back(sPart);
+				}
+			}
+
+			return (iFound > 0);
```

Comment:
Instead of iFound we can use

`return !vHosts.empty();`

Reply:
Yes/No... in this case it could be as the actual fact is that the input vector is most likely (and in this case) empty.

But when just looking at the function definition, it allows a vector as input and just ADD's values regardless if the input was empty or not. That is why the `iFound` indicates if values have been added or not.

Checking on `empty()` could be false even when nothing was added, but the input wasn't empty to begin with.

It is debatable I agree, but I prefer when looking just at the function (without considering its environment and other assumption and knowledge) it is self-explanatory. In this current way, it is in my opinion.

But if it needs to be changed...

Reply:
There is no need to have a counting value in this function, speed wise, code size etc...
It is also cleaner to use 

`return !vHosts.empty();`

Reply:
done